File: src/16_Directed_graphs/topological_sorting.cpp
#include <bits/stdc++.h>
#include "../compete.hpp"
using namespace std;


const int n = 6;
vector<int> adj[n + 1] = {
  {},
  {2},     // 1
  {3},     // 2
  {6},     // 3
  {1, 5},  // 4
  {2, 3},  // 5
  {}       // 6
};
int state[n + 1] = {0};
deque<int> top_sort;


// O(V+E)
// Returns True if a cycle is detected, False otherwise.
bool dfs(int u)
{
  if (state[u] == 1) return true;
  if (state[u] == 2) return false;
  state[u] = 1;
  for (auto v : adj[u])
    if (dfs(v)) return true;
  state[u] = 2;
  top_sort.push_front(u);
  return false;
}


// O(V+E)
void solve()
{
  for (int u = 1; u <= n; ++u) {
    if (dfs(u)) {
      cout << "Cycle detected\n";
      return;
    }
  }
  cout << top_sort << '\n';
  // copy(top_sort.rbegin(), top_sort.rend(), ostream_iterator<int>(cout, " "));
  // cout << '\n';
}


int main()
{
  solve();
  // clean-up and introduce a cycle (reverse edge (5, 3))
  adj[3] = vector<int>{5, 6};
  adj[5] = vector<int>{2};
  top_sort.clear();
  memset(state, 0, sizeof(state));
  solve();
}

File: src/16_Directed_graphs/topological_sorting.py
from collections import deque

adj = [
  [],
  [2],
  [3],
  [6],
  [1, 5],
  [2, 3],
  []
]
n = len(adj) - 1
state = [0] * (n + 1)
top_sort = deque()


# O(V+E)
def dfs(u) -> bool:
  """Returns True if a cycle is detected, False otherwise."""
  if state[u] == 1:
    # cycle detected
    return True
  if state[u] == 2:
    # already processed
    return False

  # under processing
  state[u] = 1
  for v in adj[u]:
    if dfs(v):
      return True

  # fully processed
  state[u] = 2
  top_sort.appendleft(u)
  return False


# O(V+E)
def solve():
  for u in range(1, n + 1):
    if dfs(u):
      print("Cycle detected")
      return
  print(top_sort)


if __name__ == '__main__':
  solve()

  # cleanup and introduce a cycle (reverse edge (5, 3))
  top_sort.clear()
  state = [0] * len(adj)
  adj[3] = [5, 6]
  adj[5] = [2]

  solve()
